File: tools/market_map/volatility_arbiter.js
/* volatility_arbiter.js — browser twin of volatility_arbiter.py (reliability-weighted variance blend).
 * Lets the terminal compute one horizon sigma_H from physical vol components + a VR overlay + additive
 * event/jump variance, LIVE. Locked to the Python reference via va_golden.json (test_volatility_arbiter_parity.mjs). */
(function (root, factory) {
  var mod = factory();
  if (typeof module === 'object' && module.exports) module.exports = mod;
  else root.VolArbiter = mod;
})(typeof self !== 'undefined' ? self : this, function () {

  var VERSION = 'vol_arbiter_v1';

  function component(name, sigma, reliability, base_weight, available) {
    return {
      name: String(name), sigma: +sigma, reliability: +reliability,
      base_weight: (base_weight == null ? 1.0 : +base_weight),
      available: (available == null ? true : !!available)
    };
  }

  function clip(x, lo, hi) { return Math.max(lo, Math.min(hi, x)); }
  function r6(x) { return Math.round(x * 1e6) / 1e6; }

  function vrLambda(vr, nEff, minN, kmax) {
    minN = (minN == null ? 60 : minN); kmax = (kmax == null ? 0.5 : kmax);
    if (vr == null || nEff == null || nEff < minN) return 0.0;
    var samp = clip((nEff - minN) / (3.0 * minN), 0.0, 1.0);
    var dep = clip(Math.abs(vr - 1.0) / 0.5, 0.0, 1.0);
    return r6(kmax * samp * dep);
  }

  function blend(physical, opt) {
    opt = opt || {};
    var sigma_vr = (opt.sigma_vr == null ? null : +opt.sigma_vr);
    var vr_reliability = opt.vr_reliability || 0.0;
    var event_sigma = opt.event_sigma || 0.0, jump_sigma = opt.jump_sigma || 0.0;
    var floor = (opt.floor == null ? 1e-6 : opt.floor), cap = (opt.cap == null ? 10.0 : opt.cap);

    var usable = physical.filter(function (c) { return (c.available !== false) && (+c.sigma > 0); });
    if (!usable.length) throw new Error('no usable physical volatility components');

    var raw = usable.map(function (c) {
      return Math.max(c.base_weight == null ? 1.0 : c.base_weight, 0.0) * clip(c.reliability || 0.0, 0.0, 1.0);
    });
    var tot = raw.reduce(function (a, b) { return a + b; }, 0.0);
    if (tot <= 0.0) { raw = usable.map(function () { return 1.0; }); tot = usable.length; }
    var weights = raw.map(function (r) { return r / tot; });

    var sigma2 = 0.0, i, s;
    for (i = 0; i < usable.length; i++) { s = Math.max(usable[i].sigma, 1e-12); sigma2 += weights[i] * s * s; }
    var out_weights = {};
    for (i = 0; i < usable.length; i++) out_weights[usable[i].name] = r6(weights[i]);

    if (sigma_vr != null && vr_reliability > 0.0) {
      var lam = clip(+vr_reliability, 0.0, 1.0);
      var sv = Math.max(sigma_vr, floor);
      sigma2 = (1.0 - lam) * sigma2 + lam * sv * sv;
      out_weights['vr_overlay'] = r6(lam);
    }

    var ev = Math.max(+event_sigma, 0.0), jp = Math.max(+jump_sigma, 0.0);
    var sigma2_total = sigma2 + ev * ev + jp * jp;
    var sigma = Math.min(Math.max(Math.sqrt(sigma2_total), floor), cap);

    var mean_rel = 0.0;
    for (i = 0; i < usable.length; i++) mean_rel += weights[i] * (usable[i].reliability || 0.0);
    var comps = {};
    for (i = 0; i < usable.length; i++) comps[usable[i].name] = usable[i].sigma;
    comps.event_sigma = ev; comps.jump_sigma = jp;

    return {
      sigma: sigma, sigma2: sigma2_total, weights: out_weights,
      reliability: r6(mean_rel), components: comps, version: VERSION
    };
  }

  return { component: component, blend: blend, vrLambda: vrLambda, VERSION: VERSION };
});

File: tools/market_map/volatility_arbiter.py
"""volatility_arbiter.py — reliability-weighted volatility blend.

Implements the "Scale upgrade" from the Production-Ready Integration report: instead of a single
sqrt-of-time or single-estimator sigma, produce ONE horizon sigma_H by blending physical variance
components (HV_H, EWMA/GARCH, HARQ, realized-vol) in VARIANCE space, weighted by base priority x
availability x a [0,1] reliability score, then optionally shrink toward a variance-ratio overlay
(diagnostic, not oracle), then ADD scheduled-event and jump/EVT tail variance, and take a single
square root at the end. Emits named weights + a first-class reliability score so downstream
calibration and the UI can see *why* the scale is what it is.

    w_i          = base_weight_i * availability_i * clip(reliability_i, 0, 1),  normalized
    sigma2_phys  = sum_i w_i * sigma_i^2
    sigma2_blend = (1 - lambda_VR) * sigma2_phys + lambda_VR * sigma_VR^2      (overlay, lambda in [0,1])
    sigma2_total = sigma2_blend + sigma_event^2 + sigma_jump^2
    sigma_H      = clip(sqrt(sigma2_total), floor, cap)

Pure-stdlib port of the report's numpy reference (this repo's engines are keyless / numpy-free).
Never averages sigmas directly — averaging standard deviations understates blended variance.
"""
import math

VERSION = "vol_arbiter_v1"


def component(name, sigma, reliability, base_weight=1.0, available=True):
    """A physical volatility estimate + its credibility. sigma and reliability are per-horizon."""
    return {"name": str(name), "sigma": float(sigma), "reliability": float(reliability),
            "base_weight": float(base_weight), "available": bool(available)}


def _clip(x, lo, hi):
    return max(lo, min(hi, x))


def vr_lambda(vr, n_eff, min_n=60, kmax=0.5):
    """Credibility in [0, kmax] of a variance-ratio overlay (the report treats VR as a DIAGNOSTIC,
    not an oracle). Rises with (overlap-aware) sample size and engages more as VR departs from the
    linear-variance null of 1. Returns 0 when thin or when VR ~ 1 (nothing to correct)."""
    if vr is None or n_eff is None or n_eff < min_n:
        return 0.0
    samp = _clip((n_eff - min_n) / (3.0 * min_n), 0.0, 1.0)   # 0 at min_n, 1 at 4*min_n
    dep = _clip(abs(vr - 1.0) / 0.5, 0.0, 1.0)                # |VR-1| >= 0.5 -> full departure
    return round(kmax * samp * dep, 6)


def blend(physical, sigma_vr=None, vr_reliability=0.0, event_sigma=0.0, jump_sigma=0.0,
          floor=1e-6, cap=10.0):
    """Blend variance components conservatively into one sigma_H. `physical` is a list of component()
    dicts. Returns {sigma, sigma2, weights, reliability, components, version}."""
    usable = [c for c in physical if c.get("available", True) and c.get("sigma", 0.0) > 0.0]
    if not usable:
        raise ValueError("no usable physical volatility components")

    raw = [max(c.get("base_weight", 1.0), 0.0) * _clip(c.get("reliability", 0.0), 0.0, 1.0)
           for c in usable]
    tot = sum(raw)
    if tot <= 0.0:                       # all-zero credibility -> fall back to equal weights
        raw = [1.0] * len(usable); tot = float(len(usable))
    weights = [r / tot for r in raw]

    sigma2 = sum(w * (max(c["sigma"], 1e-12) ** 2) for w, c in zip(weights, usable))
    out_weights = {c["name"]: round(w, 6) for w, c in zip(weights, usable)}

    if sigma_vr is not None and vr_reliability and vr_reliability > 0.0:
        lam = _clip(float(vr_reliability), 0.0, 1.0)
        sigma2 = (1.0 - lam) * sigma2 + lam * (max(float(sigma_vr), floor) ** 2)
        out_weights["vr_overlay"] = round(lam, 6)

    ev = max(float(event_sigma), 0.0)
    jp = max(float(jump_sigma), 0.0)
    sigma2_total = sigma2 + ev * ev + jp * jp
    sigma = min(max(math.sqrt(sigma2_total), floor), cap)

    mean_rel = sum(w * c.get("reliability", 0.0) for w, c in zip(weights, usable))
    comps = {c["name"]: c["sigma"] for c in usable}
    comps["event_sigma"] = ev
    comps["jump_sigma"] = jp
    return {"sigma": sigma, "sigma2": sigma2_total, "weights": out_weights,
            "reliability": round(mean_rel, 6), "components": comps, "version": VERSION}
